File: week1/problemSet1/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long card;
    do
    {
        card = get_long("Card no: ");
    }
    while (card < 0);

    long workingCC = card;
    int sum1 = 0;
    int sum2 = 0;

    //All the digits not being multiplied by two ...
    while (workingCC > 0)
    {
        int lastDigit = workingCC % 10;
        sum1 += lastDigit;
        workingCC = workingCC / 100;
    }
    // All the digits that are multiplied by two ...

    workingCC = card/10;    //Divide by 10 to start to get every other number starting from the second last
    while (workingCC > 0)
    {
        int lastDigit = workingCC % 10;
        int timesTwo = lastDigit * 2;
        // The last (timeTwo % 10 will give the last digit of any 2 digit product and (timesTwo/10) will resturn the first digit of any integer (due to truncating the decimal))
        sum2 = sum2 + (timesTwo % 10) + (timesTwo/10);
        workingCC = workingCC / 100;
    }
    //Checks to see if the total of the two above sums = 0 as per Luhn algorithim
    int isZero = (sum1 + sum2) % 10;
    //Re init workingCC to be the card number and determine the length of the card
    workingCC = card;
    int counter = 0;
    while (workingCC != 0) {
        workingCC = workingCC / 10;
        counter++;
    }
    long divisor = 10;
    for(int i = 0; i < counter - 2; i++)
    {
        divisor*=10;
    }
    int firstDigit = card / divisor;
    int firstTwoDigits = card / (divisor / 10);

    if (isZero != 0)
    {
        printf("INVALID\n");
    }
    else if ((counter == 13 || counter == 16) && firstDigit == 4)
    {
        printf("VISA\n");
    }
    else if (counter == 16 && (firstTwoDigits == 51 || firstTwoDigits == 52 || firstTwoDigits == 53 || firstTwoDigits == 54 || firstTwoDigits == 55))
    {
        printf("MASTERCARD\n");
    }
    else if (counter == 15 && (firstTwoDigits == 34 || firstTwoDigits == 37))
    {
        printf("AMEX\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

//Passed 15/15 tests and got 90% on the styling of the code

File: week6/PSet6/credit.py
# TODO

#input validation
while True:
    try:
        card_number = int(input("Card No: "))
    except ValueError:
        continue

    if card_number > 0:
        break

#create an array from the inputted credit card number
credit_array = [int(x) for x in str(card_number)]

sum1 = 0
sum2 = 0


for num in range(len(credit_array) - 2, -1, -2):
    if credit_array[num] * 2 < 10:
        sum1 += credit_array[num] * 2
    else:
        temp = [int(x) for x in str(credit_array[num] * 2)]
        for num in range(0, len(temp), 1):
            sum1 += temp[num]

for num in range(len(credit_array) - 1, -1, -2):
    sum2 += credit_array[num]

firstTwoDigits = int(str(credit_array[0]) + str(credit_array[1]))


if (sum1 + sum2) % 10 != 0:
    print("INVALID")
elif (len(credit_array) == 13 or len(credit_array) == 16) and credit_array[0] == 4:
    print("VISA")
elif len(credit_array) == 16 and (50 < firstTwoDigits < 56):
    print("MASTERCARD")
elif len(credit_array) == 15 and (firstTwoDigits == 34 or firstTwoDigits == 37):
    print("AMEX")
else:
    print("INVALID")

# Passed 14/14 tests and 95% styling